File: leetcode/1475. Final Prices With a Special Discount in a Shop.py
from typing import List

class Solution:
    def finalPrices(self, prices: List[int]) -> List[int]:
        s = []
        for i, el in enumerate(prices):
            if len(s) == 0 or el > s[-1][1]:
                s.append([i, el])
            else:
                while len(s) > 0 and el<=s[-1][1]:
                    prices[s.pop()[0]] -= el
                s.append([i, el])
        return prices


s = Solution()
print(s.finalPrices([8,7,4,2,8,1,7,7,10,1]))

File: Topics/stack/1475. Final Prices With a Special Discount in a Shop.cpp
#include<bits/stdc++.h>
using namespace std;

class Solution {
public:
    vector<int> finalPrices(vector<int>& p) {
        int n = p.size();
        stack<int> st;
        vector<int> ans(n);
        for (int i = n-1; i >= 0 ; i--) {
            while(!st.empty() && st.top() > p[i]) st.pop();
            if (!st.empty())  p[i]-=st.top();
            st.push(p[i]);
        }
        return p;
    }
};